File: Visual Studio 2010/VBWPAzureVideoStory/VideoStoryCreator/ViewModels/PhotoViewModel.vb
'********************************* 模块头 *********************************\
' 模块名: PhotoViewModel.vb
' 项目名: VideoStoryCreator
' 版权(c) Microsoft Corporation.
' 
' 给予通信类的基类.
' 几种ViewMode关联到phone继承的类.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'***************************************************************************/



Imports System.ComponentModel
Imports System.Windows.Media.Imaging
Imports System.IO

Public Class PhotoViewModel
    Implements INotifyPropertyChanged
    Private _name As String
    Public Property Name() As String
        Get
            Return Me._name
        End Get
        Set(value As String)
            If Me._name <> value Then
                Me._name = value
                Me.NotifyPropertyChange("Name")
            End If
        End Set
    End Property

    ' MediaStream不必支持提示变更事件.
    Public Property MediaStream() As Stream
        Get
            Return m_MediaStream
        End Get
        Set(value As Stream)
            m_MediaStream = value
        End Set
    End Property
    Private m_MediaStream As Stream

    Private _transitionDuration As Integer = 2
    Public Property TransitionDuration() As Integer
        Get
            Return Me._transitionDuration
        End Get
        Set(value As Integer)
            If Me._transitionDuration <> value Then
                Me._transitionDuration = value
                Me.NotifyPropertyChange("TransitionDuration")
            End If
        End Set
    End Property

    Private _photoDuration As Integer = 5
    Public Property PhotoDuration() As Integer
        Get
            Return Me._photoDuration
        End Get
        Set(value As Integer)
            If Me._photoDuration <> value Then
                Me._photoDuration = value
                Me.NotifyPropertyChange("PhotoDuration")
            End If
        End Set
    End Property

    Private _imageSource As BitmapImage
    Public ReadOnly Property ImageSource() As BitmapImage
        Get
            If Me.MediaStream Is Nothing Then
                Return Nothing
            End If

            If Me._imageSource Is Nothing Then
                Me._imageSource = New BitmapImage()
                Me._imageSource.SetSource(Me.MediaStream)
            End If

            Return Me._imageSource
        End Get
    End Property

    ''' <summary>
    ''' 使用相片的名称进行比较
    ''' </summary>
    Public Overrides Function Equals(obj As Object) As Boolean
        If TypeOf obj Is PhotoViewModel Then
            Return Me.Name.Equals(DirectCast(obj, PhotoViewModel).Name)
        End If
        Return False
    End Function

    Public Overrides Function GetHashCode() As Integer
        Return Me.Name.GetHashCode()
    End Function

    Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

    Protected Sub NotifyPropertyChange(propertyName As String)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
    End Sub
End Class

File: Visual Studio 2010/CSWPAzureVideoStory/VideoStoryCreator/ViewModels/PhotoViewModel.cs
/********************************* 模块头 *********************************\
* 模块名: PhotoViewModel.cs
* 项目名: VideoStoryCreator
* 版权(c) Microsoft Corporation.
* 
* 给予通信类的基类.
* 几种ViewMode关联到phone继承的类.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

using System.ComponentModel;
using System.IO;
using System.Windows.Media.Imaging;

namespace VideoStoryCreator.ViewModels
{
    public class PhotoViewModel : INotifyPropertyChanged
    {
        private string _name;
        public string Name
        {
            get { return this._name; }
            set
            {
                if (this._name != value)
                {
                    this._name = value;
                    this.NotifyPropertyChange("Name");
                }
            }
        }

        // MediaStream不必支持提示变更事件.
        public Stream MediaStream { get; set; }

        private int _transitionDuration = 2;
        public int TransitionDuration 
        {
            get { return this._transitionDuration; }
            set
            {
                if (this._transitionDuration != value)
                {
                    this._transitionDuration = value;
                    this.NotifyPropertyChange("TransitionDuration");
                }
            }
        }

        private int _photoDuration = 5;
        public int PhotoDuration
        {
            get { return this._photoDuration; }
            set
            {
                if (this._photoDuration != value)
                {
                    this._photoDuration = value;
                    this.NotifyPropertyChange("PhotoDuration");
                }
            }
        }

        private BitmapImage _imageSource;
        public BitmapImage ImageSource
        {
            get
            {
                if (this.MediaStream == null)
                {
                    return null;
                }

                if (this._imageSource == null)
                {
                    this._imageSource = new BitmapImage();
                    this._imageSource.SetSource(this.MediaStream);
                }

                return this._imageSource;
            }
        }

        /// <summary>
        /// 使用相片的名称进行比较.
        /// </summary>
        public override bool Equals(object obj)
        {
            if (obj is PhotoViewModel)
            {
                return this.Name.Equals(((PhotoViewModel)obj).Name);
            }
            return false;
        }

        public override int GetHashCode()
        {
            return this.Name.GetHashCode();
        }

        public event PropertyChangedEventHandler PropertyChanged;

        protected void NotifyPropertyChange(string propertyName)
        {
            if (this.PropertyChanged != null)
            {
                this.PropertyChanged(this.PropertyChanged, new PropertyChangedEventArgs(propertyName));
            }
        }
    }
}
